sherpa-onnx/csrc/offline-paraformer-model.cc

```diff
@@ -23,6 +23,10 @@ class OfflineParaformerModel::Impl {
     sess_opts_.SetIntraOpNumThreads(config_.num_threads);
     sess_opts_.SetInterOpNumThreads(config_.num_threads);
 
+#ifdef ENABLE_GPU
+    OrtCUDAProviderOptions cuda_options;
+    sess_opts_.AppendExecutionProvider_CUDA(cuda_options);
```

Comment:
I see many `Ort::SessionOptions` and `Ort::Allocator` handlings scattered throughout the code base. Maybe we should derive a session/allocator management class, which can be consistently applied to all models. Otherwise, we may end up having a number of such `#if` statements. Just my 2 cents. I can help to work on this. I happen to be looking at enable Intel OpenVINO these days. :-)

Reply:
Note that this is probably out of scope of this PR. Just want to share some thoughts. Thanks.

Reply:
@csukuangfj What do you think? In this way, we may not need to set session separately for every model.

Reply:
I suggest that we remove the macro and then add a new field to
https://github.com/k2-fsa/sherpa-onnx/blob/0992063de839adc41e55d0ee8a0ee1b9e4248a94/sherpa-onnx/csrc/offline-model-config.h#L15
and 
https://github.com/k2-fsa/sherpa-onnx/blob/0992063de839adc41e55d0ee8a0ee1b9e4248a94/sherpa-onnx/csrc/online-transducer-model-config.h#L13

```c++
string provider = "cpu";
```

And add the following code
```c++
if (config_.provider == "cuda") {
    OrtCUDAProviderOptions cuda_options;
    sess_opts_.AppendExecutionProvider_CUDA(cuda_options);
}
```

If we want to support other providers, such as
- `OpenVINO`
- `DirectML`
- ARM compute library
- `NNAPI`

we can add more if/else statements.

(Or we can define some enums and use a switch to replace if/else).

---

> I see many Ort::SessionOptions and Ort::Allocator handlings scattered throughout the code base. Maybe we should derive a session/allocator management class, which can be consistently applied to all models

Can we define a function
```cpp
Ort::SessionOptions GetSessionOptions(const OfflineModelConfig& config) {

}
```
and an overload
```cpp
Ort::SessionOptions GetSessionOptions(const OnlineTransducerModelConfig& config) {

}
```
to reduce duplicate code?

As for the allocator, I think the current approach is fine.